File: vol1/ch7/LeetCode99.py
class Solution:
    def recoverTree(self, root: Optional[TreeNode]) -> None:
        self.p,self.q,self.pre=None,None,None
        self.inorder(root)
        if self.p!=None and self.q!=None:
            self.p.val,self.q.val=self.q.val,self.p.val

    def inorder(self,root):
        if root==None: return
        self.inorder(root.left)
        if self.pre!=None and self.p==None and self.pre.val>root.val:
            self.p=self.pre
        if self.p!=None and self.pre.val>root.val:
            self.q=root
        self.pre=root
        self.inorder(root.right)

File: vol1/ch7/LeetCode99.cpp
class Solution {
    TreeNode*p,*q,*pre;
public:
    void recoverTree(TreeNode* root) {
        pre=p=q=NULL;
        inorder(root);
        if(p!=NULL) swap(p->val,q->val);
    }
    void inorder(TreeNode*root) {
        if(root==NULL) return;
        inorder(root->left);
        if(pre!=NULL && p==NULL && pre->val>root->val)
            p=pre;										//p为空且找到前面的逆序结点pre，则置p为pre
        if(p!=NULL && pre->val>root->val)
            q=root;									//p不为空且当前结点逆序，则置q为root
        pre=root;
        inorder(root->right);
    }
};

